### PR TITLE
Making cache tags more efficient by rewriting the API.

### PR DESCRIPTION
**Pull request by Nyholm**
I've started with the tagging, common and the redis adapter. Is this a path we want to follow?

The new API will look like this: 
```php
$item = $pool->getItem('article_40');
$item->set($post);
$item->setTags(['author_10', 'category_php']);
$pool->save($item);

// Someone requests that post
$item = $pool->getItem('article_40');

// I want to clear all post by author
$pool->clearTags(['author_10']);
```

Ping @aequasi and @andrerom. What are your thoughts?

Related to https://github.com/php-cache/issues/issues/21

TODO: 
* [x] Update the integration tests
* [x] Rewrite other adapters
* [ ] Update documentation

### CODE REVIEW

File: src/Taggable/TaggablePoolTrait.php [hunk 1]
```diff
@@ -21,148 +21,69 @@
...
+     * @param string $value
      */
-    abstract protected function getItemWithoutGenerateCacheKey($key);
+    abstract protected function appendToList($name, $value);
 
     /**
-     * Make sure we do not use any invalid characters in the tag name.
-     * If the tag name is invalid, an InvalidArgumentException should be thrown.
-     * The actual tag name will be "tag!$name".
+     * Get an array with all the valies in the list named $name
```
Comment (andrerom) [inline src/Taggable/TaggablePoolTrait.php]: Value

**Comment by Nyholm** [comment]
This PR is now ready for review. What do you think about the implementation? 

I will of course rewrite lots of documentation for this change. 

Ping @magnusnordlander

**Comment by magnusnordlander** [comment]
Okay, so I like the interface, and the way this cleans up all the methods in AbstractCachePool. I also like the implementation in TaggablePoolTrait.

My question is this: Does it really make sense for all adapters to be taggable? Some adapters are ill suited to tags, because you cannot control LRU evictions (Memcached comes to mind).

I would suggest to only implement CachableItemPoolInterface in adapters that work well with it, and then to merge my taggable PSR-6 adapter (https://github.com/fervo/cache/commit/4390a34a80bcf82ede4a4b5c75a3d00f6322105f). That way, if you (at your own peril) want to use a non-recommended adapter for tagging, you can, but since you have to jump through this extra hoop, you hopefully know that it's not recommended.

**Comment by andrerom** [comment]
@magnusnordlander wouldn't also same approach as you used with Akamai fit for that, store tags in database to be sure they are persisted, to make sure they can be used for invalidation. Small lib for that could've useful here and also in FOSHttpCache to support cache back ends w/o tagging support.

**Comment by Nyholm** [comment]
So what adapters do you suggest we should implement tagging on?

**Comment by magnusnordlander** [comment]
@andrerom Sure, that would definitely work, and would make tagging safe in any cache. In fact, I have an idea for how to implement that really nicely in my taggable PSR-6 adapter. I'll take a look at that today.

**Comment by magnusnordlander** [comment]
@Nyholm: This is a list of what I know about your current adapters:

*Suitable for storing tags along with the tagged data:*

* Doctrine
* Filesystem
* MongoDB
* PHPArray
* Predis
* Redis

*Known to be unsuitable:*

* Memcache
* Memcached

*Unknown:*

* Apc
* Apcu

*Special cases:*

* Void (not really suitable for storing *any* data, so it's equally bad for tags :) )
* Chain (if there's at least one suitable adapter in the chain, it's suitable)

**Comment by magnusnordlander** [comment]
Right, so I've made some changes. https://github.com/Nyholm/cache-1/compare/tagging...fervo:tagging

As you can see, the adapter now takes one or two cache pools. If one passes a second cache pool, that pool is used for the tags. 

That makes it safe to use Memcache as the regular pool, as long as a suitable pool is used for the cache, like so: `$taggablePool = TaggablePSR6PoolAdapter::makeTaggable($memcachedPool, $filesystemPool);`

This makes it trivial to write an adapter that stores the tags in database, all one has to do is to make sure it implements PSR-6.

**Comment by Nyholm** [comment]
I am done with this PR. Individual tests are breaking because I made BC breaking changes in common. 
Can I have one or two reviewers on this before it is merged?

I have not added tagging to MongoDB-adapter, simply because I do not know enough MongoDB.

**Comment by magnusnordlander** [comment]
LGTM :+1: :tropical_drink:

**Comment by aequasi** [comment]
:+1: Looks far cleaner

**Comment by Nyholm** [comment]
Thank you for reviewing this.

**Pull request merged**